File: backend/app/schemas/interactions.py
# backend/app/schemas/interactions.py — Pydantic-схемы для лайков и сводки рекомендаций
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime

from app.schemas.track import TrackRead


class LikeResponse(BaseModel):
    """Ответ на операции с лайками"""
    message: str
    model_config = ConfigDict(from_attributes=True)


class LikedTrackItem(BaseModel):
    """Лайк с полными данными трека (для страницы «Понравившиеся», в т.ч. недоступные)."""
    liked_at: datetime
    track: TrackRead
    model_config = ConfigDict(from_attributes=True)


class UserInteractionsSummary(BaseModel):
    """
    Сводка по взаимодействиям пользователя (для рекомендаций).
    Поле total_listens оставлено для обратной совместимости API и всегда 0.
    """
    total_likes: int
    total_listens: int = Field(default=0, description="Не используется; история прослушиваний отключена")
    favorite_genres: list[str]
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/interactions.ts
// frontend/src/api/interactions.ts — API-клиент для лайков и сводки рекомендаций
import client from './client';
import type { Track } from '../types/track';

export interface LikeResponse {
  message: string;
}

/** Элемент списка /interactions/likes с полными данными трека */
export interface LikedTrackItem {
  liked_at: string;
  track: Track;
}

export interface InteractionsSummary {
  total_likes: number;
  /** Всегда 0: история прослушиваний на бэкенде отключена */
  total_listens: number;
  favorite_genres: string[];
}

export const interactionsApi = {
  /** Добавить лайк к треку */
  async likeTrack(trackId: number): Promise<LikeResponse> {
    const response = await client.post<LikeResponse>(`/interactions/like/${trackId}`);
    return response.data;
  },

  /** Удалить лайк с трека */
  async unlikeTrack(trackId: number): Promise<LikeResponse> {
    const response = await client.delete<LikeResponse>(`/interactions/like/${trackId}`);
    return response.data;
  },

  /** Список лайков с полными данными треков */
  async getLikedTracks(limit: number = 200): Promise<LikedTrackItem[]> {
    const response = await client.get<LikedTrackItem[]>(`/interactions/likes?limit=${limit}`);
    return response.data;
  },

  /** Сводка по взаимодействиям (рекомендации строятся по лайкам) */
  async getSummary(): Promise<InteractionsSummary> {
    const response = await client.get<InteractionsSummary>('/interactions/summary');
    return response.data;
  },

  /** Проверить, лайкнут ли трек (через список лайков) */
  async isTrackLiked(trackId: number): Promise<boolean> {
    try {
      const liked = await this.getLikedTracks(200);
      return liked.some((item) => item.track.id === trackId);
    } catch {
      return false;
    }
  },
};
